File: Overlord.Core/Overlord.Core/Models/Position3D.cs
namespace Overlord.Core.Models;

/// <summary>
/// Platform-agnostic 3D position (replaces Unity's Vector3).
/// </summary>
public class Position3D
{
    public float X { get; set; }
    public float Y { get; set; }
    public float Z { get; set; }

    public Position3D()
    {
        X = 0f;
        Y = 0f;
        Z = 0f;
    }

    public Position3D(float x, float y, float z)
    {
        X = x;
        Y = y;
        Z = z;
    }

    /// <summary>
    /// Calculates distance to another position.
    /// </summary>
    public float DistanceTo(Position3D other)
    {
        float dx = X - other.X;
        float dy = Y - other.Y;
        float dz = Z - other.Z;
        return (float)Math.Sqrt(dx * dx + dy * dy + dz * dz);
    }

    /// <summary>
    /// Calculates the angle (in radians) from origin to this position in the XZ plane.
    /// </summary>
    public float GetAngleXZ()
    {
        return (float)Math.Atan2(Z, X);
    }

    /// <summary>
    /// Calculates the radius (distance from origin) in the XZ plane.
    /// </summary>
    public float GetRadiusXZ()
    {
        return (float)Math.Sqrt(X * X + Z * Z);
    }

    /// <summary>
    /// Creates a position from polar coordinates (angle and radius in XZ plane).
    /// </summary>
    public static Position3D FromPolar(float angleRadians, float radius, float y = 0f)
    {
        return new Position3D(
            (float)Math.Cos(angleRadians) * radius,
            y,
            (float)Math.Sin(angleRadians) * radius
        );
    }

    public override string ToString()
    {
        return $"({X:F1}, {Y:F1}, {Z:F1})";
    }
}

File: Overlord.Phaser/src/core/models/Position3D.ts
/**
 * Platform-agnostic 3D position (replaces Unity's Vector3).
 */
export class Position3D {
  public x: number;
  public y: number;
  public z: number;

  constructor(x: number = 0, y: number = 0, z: number = 0) {
    this.x = x;
    this.y = y;
    this.z = z;
  }

  /**
   * Calculates distance to another position.
   */
  public distanceTo(other: Position3D): number {
    const dx = this.x - other.x;
    const dy = this.y - other.y;
    const dz = this.z - other.z;
    return Math.sqrt(dx * dx + dy * dy + dz * dz);
  }

  /**
   * Calculates the angle (in radians) from origin to this position in the XZ plane.
   */
  public getAngleXZ(): number {
    return Math.atan2(this.z, this.x);
  }

  /**
   * Calculates the radius (distance from origin) in the XZ plane.
   */
  public getRadiusXZ(): number {
    return Math.sqrt(this.x * this.x + this.z * this.z);
  }

  /**
   * Creates a position from polar coordinates (angle and radius in XZ plane).
   */
  public static fromPolar(angleRadians: number, radius: number, y: number = 0): Position3D {
    return new Position3D(
      Math.cos(angleRadians) * radius,
      y,
      Math.sin(angleRadians) * radius,
    );
  }

  /**
   * Creates a deep copy of this position.
   */
  public clone(): Position3D {
    return new Position3D(this.x, this.y, this.z);
  }

  public toString(): string {
    return `(${this.x.toFixed(1)}, ${this.y.toFixed(1)}, ${this.z.toFixed(1)})`;
  }
}
